### PR TITLE
[#166927205][Do not merge] Enable SSO uplift journey for non-GDS Google users

### PR DESCRIPTION
**Pull request by mogds**
What
----

This modifies the Google SSO uplift process so non-GDS users can enable Google SSO. It removes any GDS user validation and configures the OIDC flow so the UAA username is mapped to the Google token sub during the uplift process.

How to review
-------------

- Code review
- Deploy to your dev env

**Please do not merge**

Who can review
---------------

Not me

### CODE REVIEW

File: src/components/account/oidc.ts [hunk 1]
```diff
@@ -53,9 +54,15 @@ export default class OIDC {
         {state, response_type},
       );
 
-      const oid = tokenSet.claims[KEY_OID] as string;
+      if (providerName === 'microsoft') {
+        const oid = tokenSet.claims[KEY_OID] as string;
+        await uaa.setUserOrigin(ctx.token.userID, providerName, oid);
+      }
```
Comment (46bit) [inline src/components/account/oidc.ts]: These two `if`s might look cleaner as a `switch` statement. Personal preference though.

File: src/components/account/account.ts [hunk 2]
```diff
@@ -32,21 +40,90 @@ export async function getUseGoogleSSO(ctx: IContext, _params: IParameters): Prom
...
+  }
+
+  const oidcClient = new OIDC(
+    cfgProvided.clientID,
+    cfgProvided.clientSecret,
+    cfgProvided.discoveryURL,
+    ctx.absoluteLinkTo('account.use-google-sso-callback.get'),
+  );
+
+  if (_params.hasOwnProperty('error')) {
```
[the remark this answers is not in the document]
Comment (mogds) [reply src/components/account/account.ts]: `_params` is being used to handle any errors from the SSO provider, [for example](https://github.com/alphagov/paas-admin/blob/master/src/components/account/account.ts#L185)

[the remark this answers is not in the document]
Comment (AP-Hunt) [reply src/components/account/account.ts]: To add a little extra, `_params` represents the URL query parameters that are given when the user is redirected back to us from the provider. The errors we're checking for are defined in [section 5.2](https://tools.ietf.org/html/rfc6749#section-5.2) of the OAuth spec

[the remark this answers is not in the document]
Comment (AP-Hunt) [reply src/components/account/account.ts]: We could arguably also handle the additional errors defined in the [OIDC spec](https://openid.net/specs/openid-connect-core-1_0.html#AuthError) specifically, but I'm not sure what extra value that'd give.

**Comment by 46bit** [comment]
This should be deployed before the migration script from https://github.com/alphagov/paas-cf/pull/1994. If not, users will end up relying on matching emails to sign in (reasoning is awkward to explain), which a later story may break.

**Pull request merged**